### PR TITLE
Fix SSL/TLS hostname-not-valid error

### PR DESCRIPTION
**Pull request by ff-**
As I detailed in #1 I rewrote the `HTTPRequest` class using the native `java.net.HttpsURLConnection` because the Apache Http lib doesn't support [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication) and because starting from Android 5.1 all the `org.apache.http` classes are being [deprecated](https://developer.android.com/about/versions/android-5.1.html#http).

Not supporting SNI means that when your Zulip instance is hosted on a virtualhost the TLS auth fails even if the certificate is valid and trusted.

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@ff-, thanks for the pull request! It looks like you haven't yet signed the Dropbox CLA. Please [sign it here](https://opensource.dropbox.com/cla/) and update the thread so we can consider merging your code.

**Comment by ff-** [comment]
I signed the CLA.

File: src/com/zulip/android/HTTPRequest.java [hunk 1]
```diff
@@ -1,137 +1,150 @@
...
+                throw new IOException("Wrong HTTP method.");
+            }
+
+            if (method.equals("GET")) {
+                url = new URL(app.getServerURI() + path + "?" + getPostDataString(this.postDataParams));
             }
 
+            conn = (HttpsURLConnection) url.openConnection();
+            conn.setRequestMethod(method);
+            conn.setRequestProperty("Authorization", "Basic "
```
Comment (lfaraone) [inline src/com/zulip/android/HTTPRequest.java]: Can we switch to [`java.net.Authenticator`][1] instead here?

[1]: https://developer.android.com/reference/java/net/Authenticator.html

File: src/com/zulip/android/HTTPRequest.java [hunk 2]
```diff
@@ -1,137 +1,150 @@
...
-                            + Base64.encodeToString(authstr.getBytes(),
-                                    Base64.NO_WRAP));
             // timeout after 60 seconds in ms
-            HttpConnectionParams.setSoTimeout(request.getParams(), 60 * 1000);
+            conn.setConnectTimeout(60 * 1000);
+            conn.connect();
+
+            InputStream responseStream;
+
+            if(conn.getResponseCode() >= HttpStatus.SC_BAD_REQUEST)
```
Comment (lfaraone) [inline src/com/zulip/android/HTTPRequest.java]: We should move off `HttpStatus` if possible, since leaving it in requires us to continue using deprecated libs.

File: src/com/zulip/android/HTTPRequest.java [hunk 3]
```diff
@@ -1,137 +1,150 @@
...
+            while ((line = responseStreamReader.readLine()) != null) {
+                stringBuilder.append(line).append('\n');
+            }
+
+            responseStreamReader.close();
+            responseString = stringBuilder.toString();
+            responseStream.close();
+
+            if (conn.getResponseCode() != HttpStatus.SC_OK) {
+                Log.e("HTTP", conn.getResponseMessage());
```
Comment (lfaraone) [inline src/com/zulip/android/HTTPRequest.java]: We're logging this data twice, once here and once on 147.

File: src/com/zulip/android/HTTPRequest.java [hunk 4]
```diff
@@ -1,137 +1,150 @@
...
-        }
+        return result.toString();
     }
 
     String execute(String method, String path) throws IOException {
-        AndroidHttpClient httpclient = AndroidHttpClient.newInstance(app
-                .getUserAgent());
-
-        String responseString = null;
+        String responseString = "<empty>";
```
Comment (lfaraone) [inline src/com/zulip/android/HTTPRequest.java]: Why is this changed here from `null`?

**Comment by niftynei** [comment]
@ff-  Since we moved over to OKHttp, this patch is largely obsolete.  If it's still a problem, please open an issue. 

Thanks for your work on this.
